src/lib/rules/cloudinary/cloudinary.ts

```diff
@@ -0,0 +1,188 @@
+         */
+        const cloudinary = require('cloudinary');
+        const uploads: Array<Promise<cloudinaryResult>> = [];
+        let configured = false;
+        let sizeThreshold = 0;
+
+        /* eslint-disable camelcase */
+
+        /** Sends the image to cloudinary to identify optimizations on size and format. */
+        const processImage = (data: IFetchEnd): Promise<cloudinaryResult> => {
+
+            /*
+             * Using the md5 hash of the file is the recommended way to avoid duplicates
+             * https://support.cloudinary.com/hc/en-us/articles/208158309-How-can-I-completely-prevent-the-existence-of-image-duplications-on-my-Cloudinary-account-
+             */
+            const hash = crypto
+                .createHash('md5')
+                .update(data.response.body.rawContent)
+                .digest('hex');
+
+            return new Promise((resolve, reject) => {
+                return cloudinary.v2.uploader.upload_stream(
+                    { crop: 'limit', public_id: hash, quality: 'auto' },
+                    (error, result) => {
+                        if (error) {
+                            logger.error(`Error processing image ${cutString(data.resource)} with cloudinary`);
+                            logger.error(error);
+
+                            return reject(error);
+                        }
+
+                        result.originalBytes = data.response.body.rawContent.length;
+                        result.originalUrl = data.resource;
+                        result.element = data.element;
+
+                        return resolve(result);
+                    })
+                    .end(data.response.body.rawContent);
+            });
+        };
+
+        /** Detects if there is a valid cloudinary configuration. */
+        const isConfigured = () => {
+            const cloudinaryUrl = process.env.CLOUDINARY_URL; // eslint-disable-line no-process-env
+            const { apiKey, apiSecret, cloudName, threshold } = context.ruleOptions;
+
+            if (threshold) {
+                sizeThreshold = threshold;
+            }
+
+            if (cloudinaryUrl) {
+                return true;
+            }
+
+            if (!apiKey || !apiSecret || !cloudName) {
+                logger.error('No configuration found for cloudinary');
+
+                return false;
+            }
+
+            cloudinary.v2.config({
+                api_key: apiKey,
+                api_secret: apiSecret,
+                cloud_name: cloudName
+            });
+
+            return true;
+        };
+        /* eslint-enable camelcase */
+
+        /** Analyzes the response if it's an image. */
+        const analyzeImage = (fetchEnd: IFetchEnd) => {
+            if (!configured) {
+                return;
+            }
+
+            const { response } = fetchEnd;
+
+            try {
+                // TODO: Find a better way than doing this to detect if it's an image
```

Comment:
What image formats does Cloudinary support?

Reply:
Anything? I was thinking more about all the mimetype calculations that we do in other rules for any type. We should probably do something generic for everythibg but I don't think k it hold be addressed in this PR.

Reply:
> I was thinking more about all the mimetype calculations that we do in other rules for any type.

Ah, ok.